### PR TITLE
Introducing ContentItem.DisplayText

### PR DESCRIPTION
**Pull request by sebastienros**
This introduces `ContentItem.DisplayText` and makes `TitlePat.Title` obsolete. The idea is that we always have issues with `TitlePart` being optional and not being able to use it to order items. It's also an issue to search for content items, most often requiring Lucene to be configured.

TitlePart is kept for the time being to prevent breaking existing sites, but it should be used only to display the editor and not the title itself. Template can use `ContentItem.DisplayText` directly without using the metadata anymore.

### CODE REVIEW

**Comment by Jetski5822** [comment]
So content items now have a notion of display??

**Comment by sebastienros** [comment]
See it as a ToString()

**Comment by Jetski5822** [comment]
Okay - I like

**Comment by sebastienros** [comment]
😅

**Comment by sebastienros** [comment]
@Jetski5822 The thing that decided me is that @jameshawkins is working on Content Picker. And how do you pick content naturally ... using it's title (Blog, About, ...) So we needed, like in O1, to have Lucene correctly configured, which is kind of a hassle just to find a content item. Or to be dependent on Title directly, which is bad too. And it reminded me of the same issue as we needed CommonPart on all content items in O1 and how we fixed it in OC by making it a default.

**Comment by sebastienros** [comment]
Also, in almost all templates we want to display the title, we need to browse for Title Part or to use the metadata. That's simpler now too. And we only need to add TitlePart to be able to edit the display text now.

File: src/OrchardCore.Themes/TheAgencyTheme/Recipes/agency.recipe.json [hunk 1]
```diff
@@ -1253,6 +1256,8 @@
                 "ContentItemId": "[js:uuid()]",
                 "ContentItemVersionId": "[js:uuid()]",
                 "ContentType": "Service",
+                "DisplayText": "Responsive Design",
+                "DisplayText": "Responsive Design",
```
Comment (sebastienros) [inline src/OrchardCore.Themes/TheAgencyTheme/Recipes/agency.recipe.json]: Remove

File: src/OrchardCore/OrchardCore.ContentManagement/Records/Migrations.cs [hunk 2]
```diff
@@ -43,5 +48,14 @@ public int UpdateFrom1()
 
             return 2;
         }
+
+        public int UpdateFrom2()
+        {
+            SchemaBuilder.AlterTable(nameof(ContentItemIndex), table => table
+                .CreateIndex("IDX_ContentItemIndex_DisplayText", "DisplayText")
+            );
+
```
Comment (jameshawkins) [inline src/OrchardCore/OrchardCore.ContentManagement/Records/Migrations.cs]: Should you add the new column here too?

**Comment by sebastienros** [comment]
Updated

**Comment by agriffard** [comment]
@SebastienRos So, if I want to migrate an existing OC application made before this DisplayText change, I need to add a `"DisplayText": "..."` equivalent to the `Title` of the `TitlePart` on each content item?

**Comment by sebastienros** [comment]
Re-save each content item.

**Comment by awyl** [comment]
I got this, so you can try to run, no guarantee though, It works for me.  And it loops through all the properties, so even those in bag parts will get the DisplayText.

```csharp
    public class MigrationController : Controller
    {
        private readonly ISession _session;

        public MigrationController(ISession session)
        {
            _session = session;
        }

        public async Task<IActionResult> DisplayText()
        {
            var query = await _session.Query<ContentItem>().ListAsync();

            var sb = new StringBuilder();

            foreach (var ci in query)
            {
                if (!ci.Latest) continue;

                var changed = false;

                if (!ci.Has("TitlePart"))
                {
                    sb.AppendLine($"{ci.ContentItemVersionId} {ci.Content.TitlePart.Title}");
                    ci.DisplayText = ci.Content.TitlePart.Title;
                    changed = true;
                }

                changed = Apply(ci.Content) || changed;

                if (changed)
                    _session.Save(ci);
            }

            return Content(sb.ToString());
        }

        private bool Apply(JToken content, bool firstLevel = true)
        {
            var changed = false;

            // don't bother with first level
            if (!firstLevel)
            {
                if (content.Type == JTokenType.Object)
                {
                    var title = content["TitlePart"] ? ["Title"]?.Value<string>();
                    if (!String.IsNullOrWhiteSpace(title))
                    {
                        content["DisplayText"] = title;
                        changed = true;
                    }
                }
            }

            foreach (var token in content)
            {
                changed = Apply(token, false) || changed;
            }

            return changed;
        }
    }
```

**Comment by agriffard** [comment]
I cannot resave some content items because when I try to edit in admin a blog (Ex: OC app created with TheBlogTheme recipe), there is an error so the content item cannot be displayed :

```
ArgumentNullException: Value cannot be null.
Parameter name: value
System.Text.Encodings.Web.TextEncoder.Encode(string value)


OrchardCore.DisplayManagement.Razor.RazorPage<TModel>.RenderTitleSegments(string segment, string position, IHtmlContent separator) in RazorPage.cs
AspNetCore.__Modules_OrchardCore_Contents_Views_Content_DetailAdmin.ExecuteAsync() in Content.DetailAdmin.cshtml
<h1>@RenderTitleSegments(contentItem.DisplayText)</h1>
```

**Comment by agriffard** [comment]
Check if segment is null in RenderTitleSegments. Done in #2393

**Pull request merged**